### PR TITLE
Bug 5588683: If two modules reference the same volume name, edge agent gets into a crash loop

### PR DESCRIPTION
**Pull request by CindyXing**
Currently customers can specify volume mount through the createoption; and can also add k8s style PV/storageclass in the extended feature section. The -v and k8s can both be specified or either of them. 

For above, we create PVCs and then let k8s to schedule the right PV. PV will be created in case of stroageclass or the use what customer specified. 

The root cause of the crash is that customer specify two modules mouting to the same volume. Our code creates PVC using the mount.source. When adding to the dictionary of the PVC list, the name crashes. 

The change is to 
- Include module name in PVC to prevent conflict
- Make sure using the PV name customer specified otherwise use mount.source
- Allow volume share
- mount to the name customer specified.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 1]
```diff
@@ -301,14 +301,27 @@ static string ToProxyLogLevel(LogEventLevel level)
...
-        }
+            string name = KubeUtils.SanitizeK8sValue(mount.Source);
+            string pvcName = KubeUtils.SanitizeK8sValue(module.Name) + name;
 
-        bool ShouldUsePvc() => this.persistentVolumeName.HasValue || this.storageClassName.HasValue;
+            // PVC name will be modulename + volume name
+            // Volume name will be customer defined name or modulename + mount.source
+            if (this.persistentVolumeName.HasValue)
+            {
+                this.persistentVolumeName.ForEach(volumeName => name = volumeName);
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: This re-assigning name is a bit confusing. It's can be easier in this way:
```csharp
return V1Volume(this.presistentVolumeName.GetOrElse(name), ....
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 2]
```diff
@@ -301,14 +301,27 @@ static string ToProxyLogLevel(LogEventLevel level)
...
+                return new V1Volume(name, persistentVolumeClaim: new V1PersistentVolumeClaimVolumeSource(pvcName, mount.ReadOnly));
+            }
+            else if (this.storageClassName.HasValue)
+            {
+                return new V1Volume(pvcName, persistentVolumeClaim: new V1PersistentVolumeClaimVolumeSource(pvcName, mount.ReadOnly));
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: It use to be `name` that was used for volume name, now it is changed to `pvcName`. Could you please explain what is the reason for this change?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 3]
```diff
@@ -301,14 +301,27 @@ static string ToProxyLogLevel(LogEventLevel level)
...
+            }
+            else
+            {
+                return new V1Volume(name, emptyDir: new V1EmptyDirVolumeSource());
+            }
+
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: nit: redundant blank line

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 4]
```diff
@@ -301,14 +301,27 @@ static string ToProxyLogLevel(LogEventLevel level)
...
-        V1Volume GetVolume(Mount mount)
+        V1Volume GetVolume(KubernetesModule module, Mount mount)
         {
-            string name = KubeUtils.SanitizeDNSValue(mount.Source);
-            return this.ShouldUsePvc()
-                ? new V1Volume(name, persistentVolumeClaim: new V1PersistentVolumeClaimVolumeSource(name, mount.ReadOnly))
-                : new V1Volume(name, emptyDir: new V1EmptyDirVolumeSource());
-        }
+            string name = KubeUtils.SanitizeK8sValue(mount.Source);
+            string pvcName = KubeUtils.SanitizeK8sValue(module.Name) + name;
```
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: Like the idea. Thankd Denis. 

My understanding volume share occurs only when the volume (PV) exists, i.e. customers have the volume created manually. For that, this PR didn't change the behavior. 

In case of storageclass, first we crate PVC and then PV, where PVC and PV will have the same name. If customer doesn't specify k8s attributes, the emptydir will be created. So there is no volume to share.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs [hunk 5]
```diff
@@ -41,9 +41,9 @@ bool ShouldCreatePvc(Mount mount)
             return this.storageClassName.HasValue || this.persistentVolumeName.HasValue;
         }
 
-        V1PersistentVolumeClaim ExtractPvc(Mount mount, IDictionary<string, string> labels)
+        V1PersistentVolumeClaim ExtractPvc(string moduleName, Mount mount, IDictionary<string, string> labels)
         {
-            string name = KubeUtils.SanitizeK8sValue(mount.Source);
+            string name = KubeUtils.SanitizeK8sValue(moduleName) + KubeUtils.SanitizeK8sValue(mount.Source);
```
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: ok

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs (hunk 2 above)
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: When storageclass is used, there is no PV. In this case, PVC and PV are having the same name.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs (hunk 1 above)
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: It is to ensure the volume name be what customer specified and not the mount source. As they can be different.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs (hunk 2 above)
Comment (massand) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: @CindyXing  Do we need to change logic in the CollectModuleVolumes to use pvcname when querying volumes?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs [hunk 6]
```diff
@@ -41,9 +41,9 @@ bool ShouldCreatePvc(Mount mount)
             return this.storageClassName.HasValue || this.persistentVolumeName.HasValue;
         }
 
-        V1PersistentVolumeClaim ExtractPvc(Mount mount, IDictionary<string, string> labels)
+        V1PersistentVolumeClaim ExtractPvc(KubernetesModule module, Mount mount, IDictionary<string, string> labels)
         {
-            string name = KubeUtils.SanitizeK8sValue(mount.Source);
+            string name = KubernetesModule.PvcName(module, mount);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: Nit readability: I would recommend changing variable name because name is super overloaded
```suggestion
            string pvcName = KubernetesModule.PvcName(module, mount);
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs (hunk 2 above)
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: @massand, CollectModuleVolumes did take the change in GetVolume.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs (hunk 6 above)
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: Sounds good.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 7]
```diff
@@ -301,14 +301,26 @@ static string ToProxyLogLevel(LogEventLevel level)
...
-        V1Volume GetVolume(Mount mount)
+        V1Volume GetVolume(KubernetesModule module, Mount mount)
         {
-            string name = KubeUtils.SanitizeDNSValue(mount.Source);
-            return this.ShouldUsePvc()
-                ? new V1Volume(name, persistentVolumeClaim: new V1PersistentVolumeClaimVolumeSource(name, mount.ReadOnly))
-                : new V1Volume(name, emptyDir: new V1EmptyDirVolumeSource());
-        }
+            string name = KubeUtils.SanitizeK8sValue(mount.Source);
+            string pvcName = KubernetesModule.PvcName(module, mount);
```
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: The static function is KubernetesModule specific. It is coherent to be in the same file.

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesDeploymentMapperTest.cs [hunk 8]
```diff
@@ -166,7 +166,8 @@ public void PvcMappingForVolumeNameVolume()
             Assert.True(pod != null);
             var podVolume = pod.Spec.Volumes.Single(v => v.Name == "elephant");
             Assert.NotNull(podVolume.PersistentVolumeClaim);
-            Assert.Equal(module.Name + "a-volume", podVolume.PersistentVolumeClaim.ClaimName);
+
+            Assert.Equal(module.Name + "-a-volume", podVolume.PersistentVolumeClaim.ClaimName);
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesDeploymentMapperTest.cs]: Usually, it is recommended to put some precomputed value to increase the readability of the tests. And if it fails on this line it will be obvious to see what's wrong.
I'll leave it only here but can be applied to all changes below
```suggestion 
            Assert.Equal("module1-a-volume", podVolume.PersistentVolumeClaim.ClaimName);
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs [hunk 9]
```diff
@@ -82,6 +83,11 @@ public KubernetesModule(IModule module, KubernetesConfig config)
 
         public bool Equals(IModule<KubernetesConfig> other) => this.Equals(other as KubernetesModule);
 
+        public static string PvcName(KubernetesModule module, Mount mount)
```
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesModule.cs]: PVC belongs to a module and mount doesn't always need a PVC, right?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesDeploymentMapperTest.cs (hunk 8 above)
Comment (CindyXing) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesDeploymentMapperTest.cs]: En, good point. It seems that the module.Name is a common pattern in the existing test cases. Let's have a work item to improve readability.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs [hunk 10]
```diff
@@ -301,14 +301,26 @@ static string ToProxyLogLevel(LogEventLevel level)
...
-        }
+            string volumeName = KubeUtils.SanitizeK8sValue(mount.Source);
+            string pvcName = KubernetesModule.PvcName(module, mount);
 
-        bool ShouldUsePvc() => this.persistentVolumeName.HasValue || this.storageClassName.HasValue;
+            // PVC name will be modulename + volume name
+            // Volume name will be customer defined name or modulename + mount.source
+            if (this.persistentVolumeName.HasValue)
+            {
+                this.persistentVolumeName.ForEach(pvVolumeName => volumeName = pvVolumeName);
```
[the remark this answers is not in the document]
Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/deployment/KubernetesDeploymentMapper.cs]: Sounds good. Good to see we pick up the RUST best practices.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs [hunk 11]
```diff
@@ -60,14 +60,14 @@ V1PersistentVolumeClaim ExtractPvc(KubernetesModule module, Mount mount, IDictio
             // prefer persistent volume name to storage class name, if both are set.
             if (this.persistentVolumeName.HasValue)
             {
-                this.persistentVolumeName.ForEach(volumeName => persistentVolumeClaimSpec.VolumeName = volumeName);
+                persistentVolumeClaimSpec.VolumeName = this.persistentVolumeName.OrDefault();
             }
             else if (this.storageClassName.HasValue)
             {
                 this.storageClassName.ForEach(storageClass => persistentVolumeClaimSpec.StorageClassName = storageClass);
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: If we're doing persistentVolumeName.OrDefault(), we should do that for storageClassName.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs [hunk 12]
```diff
@@ -56,17 +56,18 @@ V1PersistentVolumeClaim ExtractPvc(Mount mount, IDictionary<string, string> labe
                     Requests = new Dictionary<string, ResourceQuantity>() { { "storage", new ResourceQuantity($"{this.persistentVolumeClaimSizeMb}Mi") } }
                 },
             };
+
             // prefer persistent volume name to storage class name, if both are set.
             if (this.persistentVolumeName.HasValue)
             {
-                this.persistentVolumeName.ForEach(volumeName => persistentVolumeClaimSpec.VolumeName = volumeName);
+                persistentVolumeClaimSpec.VolumeName = this.persistentVolumeName.OrDefault();
```
Comment (philipktlin) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: nit: I think you can just call OrDefault without if block.

Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: The if block is needed as the overwrite depends on the condition. We don't want the volumename to be none when no value exists.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs (hunk 6 above)
Comment (massand) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: I cannot see this change in the file at the moment. Hope you can accept suggestion prior to merge.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs (hunk 12 above)
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: The default value of "persistentVolumeClaimSpec.VolumeName" is null; that's why you can directly call this.persistentVolumeName.OrDefault(); or you should directly call this.persistentVolumeName.Value inside if block.

Comment (CindyXing) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: We prefer none than null. :)

Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/pvc/KubernetesPvcMapper.cs]: persistentVolumeClaimSpec.VolumeName is in third-party lib, not our code.
